src/array_like.rs

```diff
@@ -0,0 +1,194 @@
+use std::marker::PhantomData;
+use std::ops::Deref;
+
+use ndarray::{Array1, Dimension, Ix0, Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{intern, sync::GILOnceCell, types::PyDict, FromPyObject, Py, PyAny, PyResult};
+
+use crate::sealed::Sealed;
+use crate::{get_array_module, Element, IntoPyArray, PyArray, PyReadonlyArray};
+
+pub trait Coerce: Sealed {
+    const VAL: bool;
+}
+
+/// Marker type to indicate that the element type received via [`PyArrayLike`] must match the specified type exactly.
+#[derive(Debug)]
+pub struct TypeMustMatch;
+
+impl Sealed for TypeMustMatch {}
+
+impl Coerce for TypeMustMatch {
+    const VAL: bool = false;
+}
+
+/// Marker type to indicate that the element type received via [`PyArrayLike`] can be cast to the specified type by NumPy's [`asarray`](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html).
+#[derive(Debug)]
+pub struct AllowTypeChange;
+
+impl Sealed for AllowTypeChange {}
+
+impl Coerce for AllowTypeChange {
+    const VAL: bool = true;
+}
+
+/// Receiver for arrays or array-like types.
+///
+/// By taking this type as the argument of a `#[pyfunction]`, one will always get access to a [`PyReadonlyArray`]
+/// that will either reference to the NumPy array originally passed into the function
+/// or a temporary one created by converting the input type into a NumPy array.
```

Comment:
How about starting from the Python side? Then maybe it's easier for users to imagine the usage. Say, "In NumPy programming in Python, it is pretty common to provide a function that takes any array-like types such as `list[float]` as its arguments. `PyArrayLike` enables the same pattern in Rust extensions."

Reply:
Thank you for the advice. I changed the text as suggested.